File: Lectures/11-Greedy-Algorithms/Huffman-Coding.py
class node:
    def __init__(self, freq, symbol, left=None, right=None, huff=''):
        self.freq = freq
        self.symbol = symbol
        self.left = left
        self.right = right
        self.huff = huff
 
def print_nodes(node, val = ''):
    new_val = val + str(node.huff)
    if (node.left):
        print_nodes(node.left, new_val)
    if (node.right):
        print_nodes(node.right, new_val)
    if (not node.left and not node.right):
        print(node.symbol, "->", new_val)
 
chars = ['a', 'b', 'c', 'd', 'e', 'f']
freq  = [ 5,   9,   12,  13,  16, 45 ]
nodes = []
for i in range(len(chars)):
    nodes.append(node(freq[i], chars[i]))
while len(nodes) > 1:
    nodes = sorted(nodes, key=lambda x: x.freq)
    left = nodes[0]
    right = nodes[1]
    left.huff = 0
    right.huff = 1
    new_node = node(left.freq+right.freq, left.symbol+right.symbol, left, right)
    nodes.remove(left)
    nodes.remove(right)
    nodes.append(new_node)
print_nodes(nodes[0])

File: Lectures/11-Greedy-Algorithms/Huffman-Coding.cpp
#include <bits/stdc++.h>
using namespace std;

struct node {
    int freq;
    string symbol;
    node* left;
    node* right;
    string huff;

    node(int freq=0, string symbol="", node* left=NULL, node* right=NULL, string huff="") {
        this->freq = freq;
        this->symbol = symbol;
        this->left = left;
        this->right = right;
        this->huff = huff;
    }
};

void print_nodes(node* curr, string val = "") {
    string new_val = val + curr->huff;
    if (curr->left)
        print_nodes(curr->left, new_val);
    if (curr->right)
        print_nodes(curr->right, new_val);
    if (!curr->left && !curr->right)
        cout << curr->symbol << " -> " << new_val << '\n';
}

int main() {
    int n = 6;
	string chars[] = {"a", "b", "c", "d", "e", "f"};
	int freq[]     = { 5,   9,   12,  13,  16, 45 };
	vector<node*> nodes;
	for (int i = 0; i < n; i++) {
        node* curr = new node(freq[i], chars[i]);
		nodes.push_back(curr);
    }
	while (nodes.size() > 1) {
        sort(nodes.begin(), nodes.end(), [](auto& x, auto& y){return x->freq < y->freq;});
		node* left = nodes[0];
		node* right = nodes[1];
		left->huff = "0";
		right->huff = "1";
		node* new_node = new node(left->freq+right->freq, left->symbol+right->symbol, left, right);
		nodes.erase(nodes.begin());
		nodes.erase(nodes.begin());
		nodes.push_back(new_node);
	}
	print_nodes(nodes[0]);
}
